kong/clustering/services/negotiation.lua

```diff
@@ -0,0 +1,270 @@
+      error("malformed service requested " .. name)
+    end
+
+    local negotiated_version = negotiate_service(name, versions)
+    services[name] = negotiated_version
+
+    log_negotiation_result(name, negotiated_version)
+  end
+
+  return services
+end
+
+
+local function register_client(cluster_data_plane_purge_delay, id, client_node)
+  local ok, err = kong.db.clustering_data_planes:upsert({ id = id, }, {
+    last_seen = ngx.time(),
+    config_hash = DECLARATIVE_EMPTY_CONFIG_HASH,
+    hostname = client_node.hostname,
+    ip = ngx.var.remote_addr,
+    version = client_node.version,
+    sync_status = client_node.sync_status,
+  }, { ttl = cluster_data_plane_purge_delay })
+
+  if not ok then
+    ngx_log(ERR, _log_prefix, "unable to update clustering data plane status: ", err)
+    return error(err)
+  end
+end
+
+function _M.init_negotiation_server(service, conf)
+  service:import("kong.services.negotiation.v1.negotiation")
+  service:set_handler("NegotiationService.NegotiateServices", function(peer, nego_req)
+    local ok, result = pcall(function()
+      local dp_id = peer.id
+      ngx_log(NOTICE, "negotiating services for DP: ", dp_id)
+      local services_requested = unwrap_services(nego_req.services_requested)
+      verify_request(nego_req)
+
+      nego_req.node.sync_status = verify_node_compatibility(nego_req.node)
+      local services = negotiate_services(services_requested)
+      register_client(conf.cluster_data_plane_purge_delay, dp_id, nego_req.node)
+
+      local nego_result = {
+        services = services
+      }
+
+      return nego_result
+    end)
+
+    if not ok then
+      ngx_log(ERR, _log_prefix, result)
+      return { error_message = result }
+    end
+
+    return result
+  end)
+end
+
+-- TODO: use event to notify other workers
+-- Currently we assume only worker 0 cares about wRPC services
+local negotiated_service
+local function init_negotiated_service_tab()
+  if not negotiated_service then
+    negotiated_service = {}
+  else
+    table_clear(negotiated_service)
+  end
+end
+
+local function set_negotiated_service(name, verison)
+  negotiated_service[name] = verison
+end
+
+local negotiation_request
+
+local function get_negotiation_request()
+  if not negotiation_request then
+    negotiation_request = {
+      node = {
+        type = "KONG",
+        version = kong.version,
+        hostname = kong.node.get_hostname(),
+      },
+      services_requested = wrap_services(asked_services),
+    }
+  end
+
+  return negotiation_request
+end
+
+function _M.negotiate(peer)
+  local response_data, err = peer:call_async("NegotiationService.NegotiateServices", get_negotiation_request())
+
+  if not response_data then
+    return nil, err
+  end
+
+  if response_data.error_message and not response_data.node then
+    return nil, response_data.error_message
+  end
+
+  init_negotiated_service_tab()
+  for name, version in pairs(response_data.services or {}) do
```

Comment:
`{}` is unwise, could we define a constant empty table?

Reply:
non-escaping allocations are sunk away.  in fact, the whole pairs and loop isn't even emitted, just as if it was all surrounded by an `if respose_data.services then  for..... end end`